### DISCUSSION TITLE
Cannot destructure property 'serviceId' of 'services.pop(...)' as it is undefined.

### DISCUSSION THREAD

**Discussion by Lorilatschki** [renovatebot/renovate#25054] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.5.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

We are using our own nuget registry (jfrog artifactory). Updating packages from this registry are working in general. However, when setting the ``LOG_LEVEL`` to "debug" we are seeing this failures in our log.

```log
DEBUG: nuget registry failure: can't get PackageBaseAddress (repository=*****)
       "err": {
         "message": "Cannot destructure property 'serviceId' of 'services.pop(...)' as it is undefined.",
         "stack": "TypeError: Cannot destructure property 'serviceId' of 'services.pop(...)' as it is undefined.\n    at getResourceUrl (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/nuget/v3.ts:70:13)\n    at Object.getReleases (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/nuget/v3.ts:163:32)\n    at getRegistryReleases (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/index.ts:77:15)\n    at mergeRegistries (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/index.ts:149:19)\n    at fetchReleases (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/index.ts:293:15)\n    at lookupUpdates (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:115:56)\n    at withLookupStats (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/workers/repository/process/fetch.ts:27:18)\n    at fetchDepUpdates (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/workers/repository/process/fetch.ts:72:30)\n    at /opt/containerbase/tools/renovate/37.5.0/node_modules/p-map/index.js:57:22"
       },
       "url": "https://*****/artifactory/api/nuget/v3/*****"
```

Does anybody know what went wrong?

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: nuget registry failure: can't get PackageBaseAddress (repository=*****)
       "err": {
         "message": "Cannot destructure property 'serviceId' of 'services.pop(...)' as it is undefined.",
         "stack": "TypeError: Cannot destructure property 'serviceId' of 'services.pop(...)' as it is undefined.\n    at getResourceUrl (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/nuget/v3.ts:70:13)\n    at Object.getReleases (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/nuget/v3.ts:163:32)\n    at getRegistryReleases (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/index.ts:77:15)\n    at mergeRegistries (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/index.ts:149:19)\n    at fetchReleases (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/modules/datasource/index.ts:293:15)\n    at lookupUpdates (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:115:56)\n    at withLookupStats (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/workers/repository/process/fetch.ts:27:18)\n    at fetchDepUpdates (/opt/containerbase/tools/renovate/37.5.0/node_modules/renovate/lib/workers/repository/process/fetch.ts:72:30)\n    at /opt/containerbase/tools/renovate/37.5.0/node_modules/p-map/index.js:57:22"
       },
       "url": "https://*****/artifactory/api/nuget/v3/*****"
```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
The error appears to be coming from the last line here: https://github.com/renovatebot/renovate/blob/0c5df8ec4600000b068534bc16c9b2c497595186/lib/modules/datasource/nuget/v3.ts#L56-L70

I'm not sure whether that means that `services.pop()` returns `undefined` or whether `serviceId` is undefined. I didn't think that the latter should trigger any such error, and I'm not sure how the former could happen.

**Reply by Lorilatschki** [reply]
@rarkins totally agree. I am not a typescript expert, even if services is empty, it should not return ``undefined`` shouldn't it?

**Reply by rarkins** [reply]
Considering that it's been mapped, filtered and sorted immediately above, I am missing how services.pop() could return undefined.

**Reply by Lorilatschki** [reply]
@rarkins is it possible to get a patched version, where the content of ``servicesIndexRaw`` is logged via ``logger.debug()``?

**Reply by rarkins** [reply]
You're welcome to submit a PR adding it to a catch block

**Reply by Lorilatschki** [reply]
Will create a PR.

**Reply by Lorilatschki** [reply]
@rarkins we found the reason for this failure on our renovate bot. We are using jfrog artifactory for our binary management system inside our company. It seems, that artifactory doesn't implement the full nuget v3 api since it doesn't offer a endpoint for [PackageBaseAddress](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource). 
Renovate uses this endpoint to determine the ``sourceUrl`` for releases.

https://github.com/renovatebot/renovate/blob/0c5df8ec4600000b068534bc16c9b2c497595186/lib/modules/datasource/nuget/v3.ts#L163-L167

I found a ticket on [jfrog](https://jfrog.atlassian.net/browse/RTFACT-26507).

**Reply by rarkins** [reply]
Thanks for the update! Is there anything more we can do on Renovate's side to work around this limitation and handle it gracefully?

**Reply by Lorilatschki** [reply]
@rarkins since this failure does **not** lead renovate to stop (_it seems that only the sourceUrl is unset_), I think there is no need to work around this limitation. Once jfrog has implemented the full v3 nuget api, I will close the discussion immediately.
